examples/manifold_gradients.dx

```diff
@@ -0,0 +1,342 @@
+represent the same rotation.
+
+' We re-use the data structures and some of the functions from the
+`quaternions.dx` example.
+
+data Quaternion = Q {x:Float & y:Float & z:Float & w:Float}
+
+-- Conversion functions from a quaternion to an array of four components.
+def quatAsVec (Q{x, y, z, w}: Quaternion) : (Fin 4 => Float) = [x, y, z, w]
+def quatFromVec ([x, y, z, w]: Fin 4 => Float) : Quaternion = Q{x, y, z, w}
+
+-- General normalization functions for arrays.
+def squareMagnitude (x:a=>Float) : Float = sum $ for i. x.i * x.i 
+def magnitude (x:a=>Float) : Float = sqrt $ squareMagnitude x
+def normalize (x:a=>Float) : a=> Float = (1.0 / (magnitude x)) .* x
+
+-- Normalization for quaternions.
+def quatNormalize (q: Quaternion) : Quaternion =
+  quatFromVec $ normalize $ quatAsVec q
+
+-- The identity quaternion, representing a no-op rotation.
+quatIdent = Q{x=0.0, y=0.0, z=0.0, w=1.0}
+
+instance Add Quaternion
+  add = \ x y. quatFromVec $ (quatAsVec x) + quatAsVec y
+  sub = \ x y. quatFromVec $ (quatAsVec x) - quatAsVec y
+  zero = quatFromVec $ zero
+
+' Scaling a quaternion does not affect the rotation it represents, but is still
+useful for numerical computations.
+
+instance VSpace Quaternion
+  scaleVec = \ s x. quatFromVec (s .* quatAsVec x)
+
+' Quaternion multiplication is composition of rotations. This formula can be
+derived from the relations that $ij = k$, $ji = -k$, $jk = i$, $kj = -i$,
+$ki = j$, $ik = -j$, $i^2 = j^2 = k^2 = -1$. These can also be used to derive
+the given expression for the quaternion inverse.
+
+instance Mul Quaternion
+  mul = \ (Q{x,y,z,w}) (Q{x=x',y=y',z=z',w=w'}). Q {
+        x = w*x' + x*w' + y*z' - z*y',
+        y = w*y' - x*z' + y*w' + z*x',
+        z = w*z' + x*y' - y*x' + z*w',
+        w = w*w' - x*x' - y*y' - z*z'
+  }
+  one = quatIdent
+
+def quatConj ((Q {x, y, z, w}):Quaternion) : Quaternion =
+  Q {x=-x, y=-y, z=-z, w=w}
+
+def quatInv (q:Quaternion) : Quaternion =
+  m = magnitude $ quatAsVec q
+  1.0 / (m * m) .* quatConj q
+
+instance Monoid Quaternion
+  mempty = one
+  mcombine = \a b. a * b
+
+
+'### Quaternions manifold structure
+
+' We now define the exponential and logarithm map for the quaternions. The
+exponential map at the identity is the well-known mapping from axis-angle to
+quaternions.  Away from the identity, we can use the fact that SO3 is a Lie
+group to transport the exponential map at the identity to the appropriate
+tangent space, by using simple right-multiplication by $q$ to associate the
+two tangent spaces.
+
+' Since we need to ensure that the exponential function is differentiable at
+zero (and similar for the logarithm), we can't use standard formulations of
+functions such as `sinc x = (sin x) / x` or
+`magnitude x = sqrt $ sum (for i. x.i * x.i)`. To get around this, we use some
+simple Taylor series expansions, noting that this will cause reduced precision,
+and running autodiff on the resulting linearizations may be inefficient.
```

Comment:
(and imprecise! see [this nice post](https://www.oxinabox.net/2021/02/08/AD-truncation-error.html) by @oxinabox)

Reply:
Good point! Added an extra comment.